Adopt more smart pointers in WebExtensionContext and WebGeolocationManagerProxy

RupinMittal opened the pull request:
#### 8864e650bf5ab2a2e90de8324002db7ba71c6f67
<pre>
Adopt more smart pointers in WebExtensionContext and WebGeolocationManagerProxy
<a href="https://bugs.webkit.org/show_bug.cgi?id=280420">https://bugs.webkit.org/show_bug.cgi?id=280420</a>
<a href="https://rdar.apple.com/136769688">rdar://136769688</a>

Reviewed by Chris Dumez, Geoffrey Garen, and Timothy Hatcher.

Smart pointer adoption as per the static analyzer.

We are now able to use WebContextSupplement::protectedProcessPool() in
WebGeolocationManagerProxy::WebGeolocationManagerProxy since 284304@main
makes it safe to ref an API object in it&apos;s constructor.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::tab const): Deleted.
(WebKit::WebExtensionAction::window const): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::getCurrentTab const):
(WebKit::WebExtensionContext::getAction):
(WebKit::WebExtensionContext::inspectorPageIdentifiers const):
(WebKit::WebExtensionContext::addExtensionTabPage):
* Source/WebKit/UIProcess/Extensions/WebExtensionAction.h:
(WebKit::WebExtensionAction::tab const):
(WebKit::WebExtensionAction::window const):
* Source/WebKit/UIProcess/WebContextSupplement.h:
(WebKit::WebContextSupplement::protectedProcessPool):
* Source/WebKit/UIProcess/WebGeolocationManagerProxy.cpp:
(WebKit::WebGeolocationManagerProxy::WebGeolocationManagerProxy):
(WebKit::WebGeolocationManagerProxy::startUpdatingWithProxy):
* Source/WebKit/WebProcess/Geolocation/GeolocationPermissionRequestManager.cpp:
(WebKit::GeolocationPermissionRequestManager::ref const):
(WebKit::GeolocationPermissionRequestManager::deref const):
* Source/WebKit/WebProcess/Geolocation/GeolocationPermissionRequestManager.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::protectedGeolocationPermissionRequestManager):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/284338@main">https://commits.webkit.org/284338@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/390feedb5f180859ec9452a89c4dabf72e90b3d8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69062 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48462 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21734 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73143 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20220 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56263 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20069 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54981 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13425 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72128 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44236 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59621 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35457 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40902 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18594 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62850 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17396 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74852 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13044 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16640 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62632 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13082 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59704 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62531 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15332 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10514 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4126 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44266 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45339 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46535 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45081 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -3195,8 +3195,9 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
     for (auto entry : m_popupPageActionMap) {
-        RefPtr tab = entry.value->tab();
-        RefPtr window = tab ? tab->window() : entry.value->window();
+        Ref webExtensionAction = entry.value;
```
Does anyone have a hypothesis as to why the static analyzer complains here? `entry` contains a `Ref<WebExtensionAction>` on the stack and so there should be no issue. I fixed it since there's probably no perf impact but am curious if anyone knows what the analyzer might be thinking.

Static analyzer failure: https://build.webkit.org/results/Apple-Sonoma-Smart-Pointer-Static-Analyzer-Build/284289@main%20(1434)/scan-build-output/StaticAnalyzer/WebKit/StaticAnalyzerReports/report-WebExtensionContextCocoa.mm-unknown-0-3846b3.html#EndPath

geoffreygaren replied:
`entry` contains a Ref, but not on the stack. `entry` is an iterator, which points inside m_popupPageActionMap. If m_popupPageActionMap were modified inside `entry.value->window()`, the Ref could be destroyed.

geoffreygaren replied:
You could consider inlining window() and/or writing it in a way that doesn't use a lambda, as a way to avoid Ref here. Not essential in this patch, but something to keep in mind.

RupinMittal replied:
Ok that makes sense! The code seems cleaner inlined and without the lambda--so I just re-uploaded with that change. Now we don't have to Ref here and at another call site that I Ref'd a few patches ago.

geoffreygaren reviewed (approved):
r=me

RupinMittal committed:
Adopt more smart pointers in WebExtensionContext and WebGeolocationManagerProxy
https://bugs.webkit.org/show_bug.cgi?id=280420
rdar://136769688

Reviewed by Chris Dumez, Geoffrey Garen, and Timothy Hatcher.

Smart pointer adoption as per the static analyzer.

We are now able to use WebContextSupplement::protectedProcessPool() in
WebGeolocationManagerProxy::WebGeolocationManagerProxy since 284304@main
makes it safe to ref an API object in it's constructor.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::tab const): Deleted.
(WebKit::WebExtensionAction::window const): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::getCurrentTab const):
(WebKit::WebExtensionContext::getAction):
(WebKit::WebExtensionContext::inspectorPageIdentifiers const):
(WebKit::WebExtensionContext::addExtensionTabPage):
* Source/WebKit/UIProcess/Extensions/WebExtensionAction.h:
(WebKit::WebExtensionAction::tab const):
(WebKit::WebExtensionAction::window const):
* Source/WebKit/UIProcess/WebContextSupplement.h:
(WebKit::WebContextSupplement::protectedProcessPool):
* Source/WebKit/UIProcess/WebGeolocationManagerProxy.cpp:
(WebKit::WebGeolocationManagerProxy::WebGeolocationManagerProxy):
(WebKit::WebGeolocationManagerProxy::startUpdatingWithProxy):
* Source/WebKit/WebProcess/Geolocation/GeolocationPermissionRequestManager.cpp:
(WebKit::GeolocationPermissionRequestManager::ref const):
(WebKit::GeolocationPermissionRequestManager::deref const):
* Source/WebKit/WebProcess/Geolocation/GeolocationPermissionRequestManager.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::protectedGeolocationPermissionRequestManager):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/284338@main
